### PR TITLE
Implement sidebar reordering settings

### PR DESCRIPTION
**Pull request by allenwq**

### CODE REVIEW

File: spec/controllers/course/admin/sidebar_settings_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      subject { get :edit, course_id: course }
+      it { is_expected.to render_template(:edit) }
+    end
+
+    describe '#update' do
+      before { allow(controller).to receive(:current_course).and_return(course) }
+      let(:sidebar_items) { controller.sidebar(:normal) }
+      let(:sample_item) { sidebar_items.sample }
+
+      context 'when the weight is great than 0' do
```
Comment (lowjoel) [inline spec/controllers/course/admin/sidebar_settings_controller_spec.rb]: greater

Comment (lowjoel) [inline spec/controllers/course/admin/sidebar_settings_controller_spec.rb]: 'when the weights are all greater than 0'

File: spec/controllers/course/admin/sidebar_settings_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+        it 'updates the weight' do
+          subject
+          saved_weight = course.reload.settings(:sidebar, sample_item[:key]).weight
+          expect(saved_weight).to eq(valid_weight)
+        end
+      end
+
+      context 'when the weight is less than 0' do
+        let(:invalid_weight) { -1 }
```
Comment (lowjoel) [inline spec/controllers/course/admin/sidebar_settings_controller_spec.rb]: you can reuse the name 'weight' but define it differently in different contexts

File: spec/features/course/admin/sidebar_settings_spec.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+        fill_in first_weight_field, with: invalid_weight
+        click_button 'submit'
+        expect(page).to have_selector('div.has-error')
+
+        fill_in first_weight_field, with: valid_weight
+        click_button 'submit'
+        expect(page).to have_field(first_weight_field, with: valid_weight)
+      end
+    end
```
Comment (lowjoel) [inline spec/features/course/admin/sidebar_settings_spec.rb]: Student, not in course?

File: app/views/course/admin/sidebar_settings/edit.html.slim [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+= simple_form_for @settings, url: course_admin_sidebar_path(current_course) do |f|
+  table.table.table-hover
+    thead
+      th = t('.name')
+      th = t('.weight')
+    tbody
+      = f.simple_fields_for :sidebar_items do |item_form|
+        tr
+          = item_form.hidden_field :id
```
Comment (lowjoel) [inline app/views/course/admin/sidebar_settings/edit.html.slim]: not valid HTML. Put this in a th or td.

File: app/controllers/components/course/users_component.rb [hunk 5]
```diff
@@ -4,6 +4,7 @@ class Course::UsersComponent
   sidebar do
     [
       {
+        key: :users,
```
Comment (lowjoel) [inline app/controllers/components/course/users_component.rb]: documentation for the key and weight and type fields?

Comment (allenwq) [inline app/controllers/components/course/users_component.rb]: Where should I document this ?

Comment (allenwq) [inline app/controllers/components/course/users_component.rb]: We already have some docs in `controllers//course/controller.rb`, but not that obvious to find.

Comment (lowjoel) [inline app/controllers/components/course/users_component.rb]: Yeah, put it there. I can't think where else to put it.

File: app/views/course/admin/sidebar_settings/edit.html.slim (hunk 4 above)
Comment (allenwq) [inline app/views/course/admin/sidebar_settings/edit.html.slim]: In the final output html it appears under tr, think it's correct ? 

And if I put this in a th/td, there will be a extra empty column in table.

File: app/controllers/components/course/users_component.rb (hunk 5 above)
Comment (allenwq) [inline app/controllers/components/course/users_component.rb]: OK, done.

File: spec/features/course/admin/sidebar_settings_spec.rb (hunk 3 above)
Comment (allenwq) [inline spec/features/course/admin/sidebar_settings_spec.rb]: For settings, I haven't handled the permissions yet.
I will do this in a separate PR after we split the abilities.

File: app/views/course/admin/sidebar_settings/edit.html.slim (hunk 4 above)
Comment (lowjoel) [inline app/views/course/admin/sidebar_settings/edit.html.slim]: you can share it with the first td.

**Comment by allenwq** [comment]
@lowjoel  done.

**Pull request merged**